src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs

```diff
@@ -0,0 +1,1277 @@
+        {
+            if (buffer.Length == 0)
+            {
+                return;
+            }
+
+            long remaining = _requestContentLengthRemaining;
+            if (remaining != -1)
+            {
+                // This HttpContent had a precomputed length, and a DATA frame was written as part of the headers. We can write directly without framing.
+
+                if (buffer.Length > _requestContentLengthRemaining)
+                {
+                    string msg = SR.net_http_content_write_larger_than_content_length;
+                    throw new IOException(msg, new HttpRequestException(msg));
+                }
+                _requestContentLengthRemaining -= buffer.Length;
+
+                if (_sendBuffer.ActiveLength != 0)
+                {
+                    // We haven't sent out headers yet, so write them together with the user's content buffer.
+
+                    // Because we have a Content-Length, we can write it in a single DATA frame.
+                    BufferFrameEnvelope(Http3FrameType.Data, remaining);
+
+                    _gatheredSendBuffer[0] = _sendBuffer.ActiveMemory;
+                    _gatheredSendBuffer[1] = buffer;
+                    await _stream.WriteAsync(_gatheredSendBuffer, cancellationToken).ConfigureAwait(false);
+
+                    _sendBuffer.Discard(_sendBuffer.ActiveLength);
+                }
+                else
+                {
+                    // Headers already sent, send just the content buffer directly.
+                    await _stream.WriteAsync(buffer, cancellationToken).ConfigureAwait(false);
+                }
+            }
+            else
+            {
+                // Variable-length content: write both a DATA frame and buffer. (and headers, which will still be in _sendBuffer if this is the first content write)
+                // It's up to the HttpContent to give us sufficiently large writes to avoid excessively small DATA frames.
+                BufferFrameEnvelope(Http3FrameType.Data, buffer.Length);
+
+                _gatheredSendBuffer[0] = _sendBuffer.ActiveMemory;
+                _gatheredSendBuffer[1] = buffer;
+                await _stream.WriteAsync(_gatheredSendBuffer, cancellationToken).ConfigureAwait(false);
+
+                _sendBuffer.Discard(_sendBuffer.ActiveLength);
+            }
+        }
+
+        private async ValueTask DrainContentLength0Frames(CancellationToken cancellationToken)
+        {
+            Http3FrameType? frameType;
+            long payloadLength;
+
+            while (true)
+            {
+                (frameType, payloadLength) = await ReadFrameEnvelopeAsync(cancellationToken).ConfigureAwait(false);
+
+                switch (frameType)
+                {
+                    case Http3FrameType.Headers:
+                        // Pick up any trailing headers.
+                        _trailingHeaders = new List<(HeaderDescriptor name, string value)>();
```

Comment:
There's only ever one trailing headers frame?

Reply:
Correct. There are no continuation frames and there can't be more than one.

Reply:
The only thing I'm not certain about is whether you can have multiple Header frames in a row. I don't believe you can, but a second set of eyes would be useful.

Reply:
https://quicwg.org/base-drafts/draft-ietf-quic-http.html#section-4.1

> An HTTP message (request or response) consists of:
> 1. the message header (see [RFC7230], Section 3.2), sent as a single HEADERS frame (see Section 7.2.2),
> 2. optionally, the payload body, if present (see [RFC7230], Section 3.3), sent as a series of DATA frames (see Section 7.2.1),
> 3. optionally, trailing headers, if present (see [RFC7230], Section 4.1.2), sent as a single HEADERS frame.

The only time we'd have multiple HEADERS frames is with informational responses (100 Continue etc.)

Reply:
Did some reading on this. H3 frames can span multiple quic frames, it's just a raw data stream, so they've done away with any frame size limits at the H3 layer. That also means they don't need continuation frames anymore, all headers come together in one frame.

Headers have a limit on the decompressed size meaning you have to load and process them before you can apply that limit. However, the compressed size will always be smaller than the decompressed size so you can limit the header frame size (& buffering) to the same value as the decompressed size limit (SETTINGS_MAX_HEADER_LIST_SIZE).

There's no need to limit the size of data frames, that data can be incrementally streamed to the app.

Reply:
> However, the compressed size will always be smaller than the decompressed size so you can limit the header frame size (& buffering) to the same value as the decompressed size limit (SETTINGS_MAX_HEADER_LIST_SIZE).

Current code does this but it isn't entirely correct if the server sends an inefficient Huffman-coded string. Should be exceptionally rare but I have a TODO to fix.

SETTINGS_MAX_HEADER_LIST_SIZE uses QPack's definition for dynamic table entry size -- 32 bytes + uncompressed length.